Server: Use lastModified() for refreshing config cache as alternate strategy

dmarteau opened the pull request:
## Description

Implements an alternate strategy for invalidating the cache:

- Implements asynchronous periodic check strategy using `QsgProject::lastModified()` for checking changed project configuration.
- Use legacy QFileSystemWatcher as default strategy
- Add `QGIS_SERVER_PROJECT_CACHE_STRATEGY` option, allow `periodic, `filesystem` or `off` for completely disabling internal cache.
- Add `QGIS_SERVER_PROJECT_CACHE_CHECK_INTERVAL` option for controlling periodic strategy
- Add tests for `QgsConfigCache` strategies.

## Motivation

This PR try to solve the following issue:

- `QgsConfigCache` use `QFileSystemWatcher` for beeing aware of project's configuration. This has majors drawbacks:
      1. Is file oriented and is not aware of changes that occur in configuration from different storage (i.e postgres,....)
      2. Does not work on NFS like mounted volumes (because of inotify limitation on Linux)

The `QsgProject::lastModified()` methods relies on the storage backend to provide the modification time of the resource. 
The cache should relies on this method for testing if the cached resources needs to be refreshed. 
 
Funded by 3Liz https://3liz.com

dmarteau committed:
Implements alternate project's cache strategies

    - FileSystem cache strategy
    - Periodic cache strategy
    - Null cache strategy

m-kuhn commented:
Thanks for tackling this.

How is this behaving with projects that are rather slow to load compared to before?
I.e. I suspect the first request now pays the price for reloading (and letting a user wait potentially for many seconds) whereas before this was done asynchronously at the moment the file was written (if notify support was available).

elpaso reviewed (changes requested):
Sorry but in its current state this is a no go for many server deployments, I recommend you to make this an opt-in configurable feature.

m-kuhn commented:
For reference, we often use this: https://github.com/opengisch/qgisserver-invalidate-cache combined with a cron job to check the modified state of nfs projects and trigger a cache invalidation

dmarteau commented:
@m-kuhn 
> How is this behaving with projects that are rather slow to load compared to before?

For my experience, this has small impact:

- If you perform an update of your project during a high traffic then you will experience latency whenever your are asynchronous or not. This will impact one request amongst many and thus will have little effect on the overall traffic 
- It will be more sensitive on low traffic, because, statistically, it will impact a larger portion of your traffic. You may always mitigate that by sending a request just after an update to trigger the reloading.

> ...combined with a cron job to check the modified state of nfs projects and trigger a cache invalidation

Yes, we dit that also with an approriate plugin that invalidate the cache when the project was updated, by removing the project from cache explicitely.

But note that triggering a refresh in the actual state does not work if your project use another storage than file and, AMHA that's the biggest issue that need to be solved.

@elpaso 

> Sorry but in its current state this is a no go for many server deployments

Can you elaborate the problems that you expect ? 

I willingly consider to make the refresh  asynchronous in some way but I really think that the cache should handle all storages and not only files - and that the current implementation is problematic: using configuration on shared (NFS or other) volumes is a  current practice (especially when distributing qgis servers  on a cluster)

m-kuhn commented:
> @m-kuhn
> 
> > How is this behaving with projects that are rather slow to load compared to before?
> 
> For my experience, this has small impact:
> 
> * If you perform an update of your project during a high traffic then you will experience latency whenever your are asynchronous or not. This will impact one request amongst many and thus will have little effect on the overall traffic

On traffic this has a small impact. On user experience, this has a big impact.

> * It will be more sensitive on low traffic, because, statistically, it will impact a larger portion of your traffic. You may always mitigate that by sending a request just after an update to trigger the reloading.

This depends on an additional step done externally. In this case all the steps could be done from externally (as described in https://github.com/qgis/QGIS/pull/47050#issuecomment-1024857584).

> 
> > ...combined with a cron job to check the modified state of nfs projects and trigger a cache invalidation
> 
> Yes, we dit that also with an approriate plugin that invalidate the cache when the project was updated, by removing the project from cache explicitely.
> 
> But note that triggering a refresh in the actual state does not work if your project use another storage than file and, AMHA that's the biggest issue that need to be solved.

Is it not possible to invalidate a DB project and let the server reload it?

dmarteau commented:
> Is it not possible to invalidate a DB project and let the server reload it?

If you mean other than with a direct call to `QgsConfigCache::removeEntry()` then no.

m-kuhn commented:
As done here? https://github.com/opengisch/qgisserver-invalidate-cache/blob/master/invalidate_cache_filter.py#L105

dmarteau commented:
> As done here? https://github.com/opengisch/qgisserver-invalidate-cache/blob/master/invalidate_cache_filter.py#L105

Exactly. 

IMHO, I think that having to rely on external trigger is a workaround for the actual incomplete/disfunctionning implementation of the cache.

m-kuhn commented:
Cache invalidation is always a controversial topic, there is probably no "one-size-fits-all" solution here. I'd prefer some additional control as @elpaso mentioned as well.

![image](https://user-images.githubusercontent.com/588407/151656841-74c1654c-3c4e-419e-84cc-25dd26c82f80.png)

dmarteau commented:
> Cache invalidation is always a controversial topic, there is probably no "one-size-fits-all" solution here. I'd prefer some additional control as @elpaso mentioned as well.

Indeed :-)

I have no doubt that asynchronous cache refresh is a better option, but the actual implementation is broken for other storage and NFS and that's not a philosophical matter :-p

I would suggests:
- Cache testing is done asynchronously using a throttle so it may use storage metadata for invalidating the cache and work
  on shared volumes.
- Use an option to totally prevent cache refresh so that in can be handled completely externally

dmarteau commented:
@m-kuhn @elpaso I just realize that all assumption above about the increasing latency of request is totally wrong !

What the actual asynchronous notification does is  **simply remove the project from the cache** so you **always pay the cost** of reloading the project at the next request when the project is updated,  so in fact the proposed solution has - almost - no impact compared to the actual implementation.

m-kuhn commented on src/server/qgsserversettings.cpp:
```diff
@@ -333,6 +333,18 @@ void QgsServerSettings::initSettings()
                                   };
   mSettings[ sServiceUrl.envVar ] = sWmtsServiceUrl;
+
+  // the default WMTS service URL.
```
WMTS?

dmarteau replied:
Typo....

dmarteau replied:
Fixed.

dmarteau commented:
Changed to asynchronous invalidation check as optional strategy.  Using file system watcher as default.

m-kuhn commented:
> Changed to asynchronous invalidation check as optional strategy. Using file system watcher as default.

I can imagine that this order will be challenged.

dmarteau commented on src/server/qgsserversettings.h:
```diff
@@ -77,6 +77,7 @@ class SERVER_EXPORT QgsServerSettingsEnv : public QObject
       QGIS_SERVER_WCS_SERVICE_URL, //!< To set the WCS service URL if it's not present in the project. (since QGIS 3.20).
       QGIS_SERVER_WMTS_SERVICE_URL, //!< To set the WMTS service URL if it's not present in the project. (since QGIS 3.20).
       QGIS_SERVER_LANDING_PAGE_PREFIX, //! Prefix of the path component of the landing page base URL, default is empty (since QGIS 3.20).
+      QGIS_SERVER_CONFIG_CACHE_CHECK_INTERVAL, //! Set the interval for cache invalidation check, default to 0 which select the legacy File system watcher  (since QGIS 3.24).
```
Do you mean completely disabling all cache invalidation, including file system watcher ?

m-kuhn replied:
Yes, exactly.

dmarteau commented:
> I can imagine that this order will be challenged.

Considering the discussion above, I choose to stick to the actual behavior as default.

m-kuhn commented:
That makes sense, sorry I misread that.
I'd be happy to get a review from @elpaso again

Can you also update the description of this PR so it contains the latest state of available configuration options?

elpaso commented on src/server/qgsconfigcache.h:
```diff
@@ -65,20 +66,46 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+     *
+     * \since QGIS 3.24
+     */
+    void setCheckInterval( int msec );
```
I might be missing the point here, but I would like to see a separate logic to control the filesystem watcher and the new timer-based watcher.
Am I correct that given this implementation the two are mutually exclusive?
I see a use case where there is a mix of filesystem-based projects and other storages, but the timer-base approach would not work with the filesystem-based projects.

dmarteau replied:
> Am I correct that given this implementation the two are mutually exclusive?

Yes

> but the timer-base approach would not work with the filesystem-based projects.

Why do you think it wont't work ? The timer approach is  based is based on the 'QgsProject::lasteModified()` call which handle file based projects as well as other storage.

Since the timer strategy handle files the two methods will probably walk on each other steps.

elpaso replied:
> > Am I correct that given this implementation the two are mutually exclusive?
> 
> Yes
> 
> > but the timer-base approach would not work with the filesystem-based projects.
> 
> Why do you think it wont't work ? The timer approach is based is based on the 'QgsProject::lasteModified()` call which handle file based projects as well as other storage.

The lastModified call calls readProjectStorageMetadata which is not implemented for filesystem-based projects and even if it was implemented it would need to reload and parse the project's XML which is one of the slow steps here: we definitely want to avoid that unless it's implemented in a separate thread.

m-kuhn replied:
On which level should this be configurable? Per project? Path prefix?

I think the current approach suits many cases (actually I am not aware of an installation where we would mix different storage types with one server instance) and it's quite simple, which speaks for it in terms of expected problems and maintenance overhead.

My approach would be to offer some simple strategies (as done here) plus an option to disable this completely and implement dedicated strategies via custom logic (plugin).

dmarteau replied:
> The lastModified call calls readProjectStorageMetadata which is not implemented for filesystem-based projects

AFAIK, the lastModified() handle files by calling directly `QFileInfo::lastModified`: https://qgis.org/api/qgsproject_8cpp_source.html#l00686

> It would need to reload and parse the project's XML which is one of the slow step

Why it would need to do that ? this is a simple modified time check on a file.

elpaso replied:
> Why it would need to do that ? this is a simple modified time check on a file.

You are right: I thought there was a filesystem storage implementation which is not the case.

dmarteau replied:
Note that the tests ares based on files :-)

elpaso reviewed (commented):
@dmarteau thank you for the changes, we are heading in the right direction and I think that the timer-based cache invalidation will be very useful.

I only have a concern (highlighted in the code comments) about the mutual exclusivity of the two invalidation mechanisms, can you please clarify that?

dmarteau commented on src/server/qgsconfigcache.h:
```diff
@@ -65,20 +66,46 @@ class SERVER_EXPORT QgsConfigCache : public QObject
 
     QCache<QString, QDomDocument> mXmlDocumentCache;
-    QCache<QString, QgsProject> mProjectCache;
+    QCache<QString, std::pair<QDateTime, std::unique_ptr<QgsProject> > > mProjectCache;
```
AMHA this would be confusing: there is only one cache, this is not related specifically to the timer cache: it is a replacement for 'QCache<project>' that add the lastModified Time  QDateTime.

elpaso replied:
Fair point.

dmarteau commented on src/server/qgsconfigcache.cpp:
```diff
@@ -152,25 +154,83 @@ QDomDocument *QgsConfigCache::xmlDocument( const QString &filePath )
+{
+  if ( msec <= 0 )
+  {
+    qDebug() << QString( "Config cache: activating file system watcher" );
```
Done

dmarteau commented on src/server/qgsconfigcache.cpp:
```diff
@@ -152,25 +154,83 @@ QDomDocument *QgsConfigCache::xmlDocument( const QString &filePath )
+  }
+  else
+  {
+    qDebug() << QString( "Config cache: activating timed check" );
```
Done

dmarteau commented on src/server/qgsconfigcache.cpp:
```diff
@@ -14,6 +14,7 @@
  *   (at your option) any later version.                                   *
  *                                                                         *
  ***************************************************************************/
+#include <QDebug>
```
Done

dmarteau commented on src/server/qgsconfigcache.cpp:
```diff
@@ -152,25 +154,83 @@ QDomDocument *QgsConfigCache::xmlDocument( const QString &filePath )
+{
+  // QCache::keys returns a QList so it is safe
+  // to mutate while iterating
+  for ( auto path : mProjectCache.keys() )
```
Done

dmarteau commented on src/server/qgsconfigcache.h:
```diff
@@ -65,20 +66,46 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    QCache<QString, std::pair<QDateTime, std::unique_ptr<QgsProject> > > mProjectCache;
+
+    //! Timer for checking cache invalidation
+    QTimer mTimer;
```
Done

dmarteau commented on src/server/qgsconfigcache.cpp:
```diff
@@ -117,7 +117,9 @@ const QgsProject *QgsConfigCache::project( const QString &path, const QgsServerS
   }
-  return mProjectCache[ path ];
+
+  auto entry = mProjectCache[ path ];
```
Done

m-kuhn replied:
Also, please only use `auto` where it gives considerable advantages (e.g. readability) and use explicit types for simple things like `QString`, `QVariant`, ...

dmarteau commented on tests/src/python/test_qgsserver_configcache.py:
```diff
@@ -0,0 +1,94 @@
+"""QGIS Unit tests for QgsConfigCache
+
+From build dir, run: ctest -R PyQgsServerCacheManager -V
```
Done

dmarteau commented on tests/src/python/test_qgsserver_configcache.py:
```diff
@@ -0,0 +1,94 @@
+
+        # Activate timed check
+        # Next check should occurs in 3s
+        cache.setCheckInterval(3000)
```
Done

dmarteau commented:
As suggested by @m-kuhn I have added an option to disable cache completely

m-kuhn commented on python/server/auto_generated/qgsconfigcache.sip.in:
```diff
@@ -53,6 +53,21 @@ value is ``False``).
+If ``msec`` is set to a positive value then the cache will check for invalidation
+every ``msec`` milliseconds
+
+If ``msec`` is set to 0 (or negative value), which is the default, (or negative value) then the cache will use the QFileSystemWatcher for
```
repetition here

dmarteau replied:
Done

m-kuhn commented on src/server/qgsserversettings.h:
```diff
@@ -77,6 +77,8 @@ class SERVER_EXPORT QgsServerSettingsEnv : public QObject
       QGIS_SERVER_WMTS_SERVICE_URL, //!< To set the WMTS service URL if it's not present in the project. (since QGIS 3.20).
       QGIS_SERVER_LANDING_PAGE_PREFIX, //! Prefix of the path component of the landing page base URL, default is empty (since QGIS 3.20).
+      QGIS_SERVER_CONFIG_CACHE_CHECK_INTERVAL, //! Set the interval for cache invalidation check, default to 0 which select the legacy File system watcher  (since QGIS 3.24).
+      QGIS_SERVER_DISABLE_CONFIG_CACHE, //! Disable internal config cache  (since QGIS 3.24).
```
```suggestion
      QGIS_SERVER_PROJECT_CACHE_STRATEGY, //! 'filesystem', 'interval' or 'off' \since QGIS 3.24
```

Suggestion to
- control the behavior mainly via this (string/enum based)
- only use the interval configuration if this is set here ( no special value / negative ... handling)
- use the term `project cache` rather than 'config cache', I think that's more precise as term

dmarteau replied:
I would use `periodic` instead of `interval`.

dmarteau replied:
@m-kuhn @elpaso wouldn't it be better to separate the two strategies in two QgsConfigCache sub-classes ?

dmarteau replied:
Or separate the two strategies in different classes...

dmarteau commented:
@m-kuhn  The proposed implementation will disable cache completely not only invalidation (which could be hazardous), this is what you expect ?

dmarteau commented:
I

> Yes, disabled cache invalidation is what I meant.

That's dangerous, because with no cache invalidation you never update your project unless adding the appropriate alternate invalidation strategy.

m-kuhn commented:
I believe it's a totally valid admin decision to disable magic and take control.

dmarteau commented:
Pushed changes:
- Split strategy implementations in their own classes.
- Fixed disabling invalidation and invalidation only - no the cache itself - by implementing Null cache strategy.

@m-kuhn @elpaso are you ok with that ?

m-kuhn commented:
The new update on first sight will need more review. At least some things need adjustments (documentation, camelCase, protected for virtual methods) @elpaso what's your opinion?

dmarteau commented:
squashed commits.

dmarteau commented:
Dealing with 3 strategies was becoming painful to implements in one piece of code. Splitting strategies make them more readable and easier to maintain.

dmarteau commented:
Note for the protected virtuals, the requirement was:
- Not public
- Overridable
- Callable from QgsConfigCache

m-kuhn commented:
> Note for the protected virtuals, the requirement was:
> 
> * Not public
> * Overridable
> * Callable from QgsConfigCache

Nr 1 and 3 are mutually exclusive I think?


In general I prefer the Herb Sutter style: Public non virtual that calls private virtual.

m-kuhn commented on src/server/qgsconfigcache.h:
```diff
@@ -30,20 +31,59 @@
+ * \brief Abstract cache strategy for server configuration
+ * \since QGIS 3.24
+ */
+class SERVER_EXPORT QgsAbstractCacheStrategy  : public QObject
```
Why is this a QObject?
Is there any benefit in making this public api? I guess that could be moved into a private header easily.

dmarteau replied:
I was concerned about testability, so there should be a way to test strategy changes directly from `QgsConfigCache`. 

If we made these private we need to expose public api to change strategy with different parameters.

I can add method for activating strategy like `setFileSystemStrategy(...)` and make the strategy private and so on but we will loose the generericity of the approach and I'm pretty sure that these methods will stick around for a long time.

m-kuhn replied:
Why is this a QObject?

Is modifying environment variables not enough for testing purpose?

The effect of making this public is that we will not be able to modify this api any more in the future, so there is a risk this will block some required fixes/improvements in the future. That's why I think we should only make things public that are meant to be public. I.e. if we ever have a registry where someone could register customer implementations of caching strategies, this will need to be public api, but at the current point in time I don't think this is justified.

dmarteau replied:
> Is modifying environment variables not enough for testing purpose?

It looks like no:  once you get the instance you won't build able to change the strategy

> The effect of making this public is that we will not be able to modify this api any more in the future

I'm aware of that, may be we can mark that api as experimental ? This is a common practice.

m-kuhn replied:
I don't see this as experimental, it's an internal implementation detail.

dmarteau replied:
Anyway, for testability we need to be able to set the strategy in a unitary way...

m-kuhn replied:
I would prefer to use the env and completely reinitialize everything.
Otherwise at least kick it out of the python bindings.

dmarteau replied:
> I would prefer to use the env and completely reinitialize everything.

How would you do that ? Recreating a `QgServer` is not enough since `init()`, createInstance()`,  rely on all statics variables...

> Otherwise at least kick it out of the python bindings.

Maybe the only viable solution for testing, even if I found that no very convenient (need no reimplement `touch()` :-/)

m-kuhn commented on src/server/qgsconfigcache.h:
```diff
@@ -30,20 +31,59 @@
+ */
+class SERVER_EXPORT QgsAbstractCacheStrategy  : public QObject
+{
+    friend class QgsConfigCache;
```
No reason to use all the friend magic. Just make this public api. And make an internal private virtual to be implemented by subclasses.

dmarteau replied:
In the actual implementation, the strategy si responsible for connecting `QgsConfigCache` private slots methods so these must be made public if we do not use  `friends` declaration.

m-kuhn replied:
Do they need to be private?

m-kuhn commented on src/server/qgsconfigcache.h:
```diff
@@ -30,20 +31,59 @@
+
+    Q_OBJECT
+  public:
+    virtual QString ident() const = 0;
```
Can this be called `name` o(I think that would be consistent with other naming choices you did in this pr?) Or otherwise a full name like `identifier`or similar.

dmarteau replied:
Fixed,

> > Note for the protected virtuals, the requirement was:
> > 
> > * Not public
> > * Overridable
> > * Callable from QgsConfigCache
> 
> Nr 1 and 3 are mutually exclusive I think?
> 
> In general I prefer the Herb Sutter style: Public non virtual that calls private virtual.

That's why I use friendship.

m-kuhn commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +105,155 @@ class SERVER_EXPORT QgsConfigCache : public QObject
-    QgsConfigCache() SIP_FORCE;
+    /**
+     * Set the invalidation strategy for the cache
+     * The cache takes ownership of the strategy which will be deleted on cleanup
```
What does "on cleanup" mean? Remove this or clarify.

dmarteau replied:
Fixed.

m-kuhn commented on src/server/qgsconfigcache.h:
```diff
@@ -30,20 +31,59 @@
      * Returns the current instance.
      */
-    static QgsConfigCache *instance();
+    static QgsConfigCache *instance( QgsServerSettings *settings = nullptr );
```
I believe this is mixing concerns of initialisation and acquiring an instance. Better separate these two tasks.

dmarteau replied:
An instance should not be created with no strategy, because  `QgsConfigCache::instance()` is public api and users expect
to get the default strategy.

In the context  of the server inititalisation this would mean:
1- Create the instance with the default strategy (allocate an object)
2- Immediately after,  change the strategy according to the settings (destroy the previous object and reallocate another strategy
 
That's sound not very efficient - but I can live with that.

Another option would be to create another function `createInstanceWithOptions`  called from within the server `init()` only.

m-kuhn replied:
The second thing is what I was thinking of -- you could even add a `Q_ASSERT( settings )` in `instance()` to complain if someone tries to access instance before init.

dmarteau replied:
>  you could even add a Q_ASSERT( settings ) in instance() to complain if someone tries to access instance before init.

This may be considered as an API break or I'm wrong ?

m-kuhn replied:
Q_ASSERT is a debug build safety belt only, replace it with other logging mechanism if you prefer. Also I don't see that happen in the wild, this will always be initialized internally.
Crashing / excepting on a misuse of API is IMHO always preferable over UB.

dmarteau replied:
> Also I don't see that happen in the wild

As long as QgsConfigCache is not used outside the context of the server `init()`. I think that @elpaso use it in some  scripts.

dmarteau replied:
> Q_ASSERT is a debug build safety belt only, replace it with other logging mechanism

Ok, so that won't prevent instance creation, then we have to return an instance with the default strategy.

m-kuhn replied:
If we say "It is required to call init() before the first call to instance(), not doing so is undefined behavior" we do not need to create the default strategy (and we help developers working in developer mode to find this by Q_ASSERT'ing).

dmarteau replied:
>  not doing so is undefined behavior

Not really , because it always define a default strategy. There is a also log message to warn that init() has not been called and that the default behavior is used.

elpaso commented on src/server/qgsserversettings.cpp:
```diff
@@ -333,6 +333,29 @@ void QgsServerSettings::initSettings()
+                                            };
+  mSettings[ sServiceUrl.envVar ] = sConfigCacheCheckInterval;
+
+  // the default config cache check interval
```
```suggestion
  // the default config cache strategy
```

dmarteau replied:
Fixed.

elpaso commented on src/server/qgsserversettings.cpp:
```diff
@@ -333,6 +333,29 @@ void QgsServerSettings::initSettings()
+  // the default config cache check interval
+  const Setting sDisableConfigCache = { QgsServerSettingsEnv::QGIS_SERVER_PROJECT_CACHE_STRATEGY,
+                                        QgsServerSettingsEnv::DEFAULT_VALUE,
+                                        QStringLiteral( "Project's cache strategy" ),
```
maybe useful to list the possible values in the description

dmarteau replied:
Done.

elpaso commented on src/server/qgsserversettings.cpp:
```diff
@@ -615,3 +638,23 @@ QString QgsServerSettings::serviceUrl( const QString &service ) const
+       !result.compare( QLatin1String( "periodic" ) ) ||
+       !result.compare( QLatin1String( "off" ) ) )
+  {
+    QgsMessageLog::logMessage( QStringLiteral( "Invalid cache strategy, using 'filesystem' as default." ), "Server", Qgis::MessageLevel::Warning );
```
maybe useful to list the allowed values in the message.

dmarteau replied:
Done

elpaso commented on src/server/qgsserversettings.h:
```diff
@@ -77,6 +77,8 @@ class SERVER_EXPORT QgsServerSettingsEnv : public QObject
       QGIS_SERVER_WMTS_SERVICE_URL, //!< To set the WMTS service URL if it's not present in the project. (since QGIS 3.20).
       QGIS_SERVER_LANDING_PAGE_PREFIX, //! Prefix of the path component of the landing page base URL, default is empty (since QGIS 3.20).
+      QGIS_SERVER_PROJECT_CACHE_CHECK_INTERVAL, //! Set the interval for cache invalidation strategy 'interval', default to 0 which select the legacy File system watcher  (since QGIS 3.24).
+      QGIS_SERVER_PROJECT_CACHE_STRATEGY, //! Define the project cache strategy (since QGIS 3.24).
```
```suggestion
      QGIS_SERVER_PROJECT_CACHE_STRATEGY, //! Set the project cache strategy (since QGIS 3.24).
```

dmarteau replied:
Fixed.

elpaso commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +105,155 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    QString ident() const override { return QStringLiteral( "periodic" ); };
+
+    /**
+     *src/server/qgsserver.cpp Set the invalidation check interval for PeriodicStrategy
```
```suggestion
     * Sets the invalidation check interval for 'periodic' strategy
```

dmarteau replied:
Fixed

elpaso commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +105,155 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+     * If \a msec is set to a positive value then the cache will check for invalidation
+     * every \a msec milliseconds
+     *
+     * If \a msec is set to 0 (or negative value), which is the default,
```
I'm not convinced about the side-effect of setting the interval, the function doc say that setting it to 0 automatically (de)activates the 'periodic' strategy and activates the 'filesystem' strategy, I think it would be more intuitive to remove this side-effect: setting it to 0 would just effectively deactivate the periodic timer and implicitly deactivate any caching removal strategy (a log message would be probably useful in that case).

dmarteau replied:
> That setting it to 0 automatically (de)activates the 'periodic' strategy and activates the 'filesystem' strategy

This is not the case anymore, there is no side-effects between strategies parameters in the current implementation. Setting the strategy is now explicit.

I need to check the documentation.

dmarteau replied:
Fixed the documentation.

m-kuhn commented on src/server/qgsconfigcache.cpp:
```diff
@@ -152,25 +187,128 @@ QDomDocument *QgsConfigCache::xmlDocument( const QString &filePath )
+{
+  // QCache::keys returns a QList so it is safe
+  // to mutate while iterating
+  for ( const auto &path : mProjectCache.keys() )
```
take a const copy before iterating

dmarteau replied:
I'm confused, a copy wont mutate the original, why make it const ?

dmarteau replied:
Afaik, the keys() method returns already a copy so it should be safe to use it directly in the iteration.

dmarteau replied:
Done, but I'm curious to know why the previous statement was not suitable.

m-kuhn replied:
See "avoid detach" in https://www.kdab.com/goodbye-q_foreach/ for the technical background

pblottiere commented on src/server/qgsserversettings.h:
```diff
@@ -294,6 +296,27 @@ class SERVER_EXPORT QgsServerSettings
+     *
+     * \since QGIS 3.24
+     */
+    QString projectCacheStrategy() const;
```
What do you think about defining an enum for possible values instead of plain strings?

dmarteau replied:
Not sure that it will bring improvement since stragegy classes are internally  instancied directly  from settings values.

pblottiere replied:
I think that string comparisons should be avoided when possible. It's not robust and hardly readable compared to enum. BTW QGIS Server 2.X source code was very hard to maintain and buggy notably for this reason.

I won't block this PR (especially if @elpaso and @m-kuhn are ok with the current implementation) but I'll probably take care of it at one time or another.

dmarteau replied:
> I think that string comparisons should be avoided when possible

Whatever you do you can prevent string comparison when converting the settings to enum and there is nowhere else than initialisation where such comparison is done.

pblottiere replied:
it always start this way

pblottiere commented on src/server/qgsserversettings.h:
```diff
@@ -294,6 +296,27 @@ class SERVER_EXPORT QgsServerSettings
+
+    /**
+     * Returns the project's cache strategy
+     * The default value is 'filesystem', the value can be change by setting the environment
```
```suggestion
     * The default value is 'filesystem', the value can be changed by setting the environment
```

dmarteau replied:
fixed

pblottiere commented on src/server/qgsserversettings.h:
```diff
@@ -294,6 +296,27 @@ class SERVER_EXPORT QgsServerSettings
     QString serviceUrl( const QString &service ) const;
 
+    /**
+     * Returns the config cach check interval for the 'interval' strategy.
```
`Periodic` or `interval` strategy?

DelazJ replied:
BTW, cache

dmarteau replied:
'periodic'
 
fixed

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -30,6 +31,38 @@
+    virtual void entryInserted( const QString &path ) = 0;
+
+    //! Attache cache to this strategy
+    virtual void attach( QgsConfigCache *cache ) = 0;
```
Can we use an reference to `QgsConfigCache` in this case?

dmarteau replied:
Done

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+ * \ingroup server
+ * \brief Null system cache strategy for server configuration, it basically does nothing,
+ * completely disabling cache invalidation
+ * invalidation
```
Doc looks strange here.

dmarteau replied:
> Doc looks strange here.

In what matter ? It describes what the strategy does, i.e nothing (may be `it basically does nothing` is useless)

dmarteau replied:
Changed wording.

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+     * If \a msec is set to a positive value then the cache will check for invalidation
+     * every \a msec milliseconds
+     *
+     * \since QGIS 3.24
```
`\since` is not necessary

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    QString name() const override { return QStringLiteral( "periodic" ); };
+
+    /**
+     * Set the invalidation check interval for PeriodicStrategy
```
```suggestion
     * Sets the invalidation check interval for PeriodicStrategy
```

dmarteau replied:
Fixed

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+     */
+    void setCheckInterval( int msec );
+
+    //! Return the invalidation check interval
```
```suggestion
    //! Returns the invalidation check interval
```

dmarteau replied:
Fixed

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    //! Return the invalidation check interval
+    int checkInterval() const { return mInterval; }
+
+    //! Attach cache to this strategy
```
```suggestion
    //! Attaches cache to this strategy
```

dmarteau replied:
Fixed

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    //! The name of the strategy
+    QString name() const override { return QStringLiteral( "off" ); };
+
+    //! Attach cache to this strategy
```
```suggestion
    //! Attaches cache to this strategy
```

dmarteau replied:
Fixed

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    void entryRemoved( const QString &path ) override;
+
+    //! Called when an entry is inserted
+    void entryInserted( const QString &path ) override;
```
`path` documentation

dmarteau replied:
Fixed

pblottiere commented on src/server/qgsserversettings.cpp:
```diff
@@ -615,3 +638,23 @@ QString QgsServerSettings::serviceUrl( const QString &service ) const
+       !result.compare( QLatin1String( "off" ) ) )
+  {
+    QgsMessageLog::logMessage( QStringLiteral( "Invalid cache strategy, expecting 'filesystem', 'periodic' or 'off'. Using 'filesystem' as default." ), "Server", Qgis::MessageLevel::Warning );
+    result = QString( "filesystem" );
```
@m-kuhn What is preferable in this case: `QStringLiteral` or `QLatin1String`? I always forget :(.

m-kuhn replied:
```suggestion
    result = QStringLiteral( "filesystem" );
```

Quoting [Qt](https://www.qt.io/blog/2014/06/13/qt-weekly-13-qstringliteral)
> Use QStringLiteral only if a QString needs to be created

I think we are clearly creating a `QString` here

m-kuhn replied:
@dmarteau this suggestion is good to apply

dmarteau replied:
Fixed

m-kuhn commented:
Requesting re-review after the latest changes.
Please also leave a message to specify what you reviewed / how you got to the conclusion (CC @rldhont)

m-kuhn commented on src/server/qgsconfigcache.cpp:
```diff
@@ -23,27 +22,73 @@
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Project's cache not initialized, using 'filesystem' as default strategy" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Info );
     sInstance = new QgsConfigCache();
```
```suggestion
    Q_ASSERT( false )
    return nullptr;
```

m-kuhn replied:
This should crash immediately. Otherwise CI will still be green and we'll happily merge and everyone on the planet lives with a couple of strange error messages in their logs.

dmarteau replied:
Not sure with that, IHMO this is an api break, I would like @elpaso opinion on this because afaik he is a consumer of the QgsConfigCache api.

m-kuhn replied:
@elpaso are you ok if we foce require explicit initialization of the `QgsConfigCache` object prior to being able to instantiate it?
This is a direct consequence of the decision to make this a singleton.

elpaso replied:
@dmarteau we are calling `QgsConfigCache.instance().removeEntry(path)` but only after a `QgsServer()` has been created so I guess there should not be any problem, can you confirm?

dmarteau replied:
`Init()` is called at server creation so it is ok in the server context. 

But my concern is about the fact that 'QgsConfigCache::instance()` is public api and can be legitimately used outside of the context of `QgsServer` instanciation in any python script. So, requiring that `QgsConfig::initialize(...)` must be called before any call to 'QgsConfigCache::instance()` will break these scripts.  So, in the absolute, this is an api break.

m-kuhn replied:
Technically you are correct.
However, is there a real world use case where the first call to `instance()` is done from outside qgis core? If there is actually such a scenario, and some qgis server plugin does that and subsequently the qgis server core implementation calls `init()` because it has no way of knowing, what would be the desired behavior? I.e. technically speaking isn't that already an API break already?

FWIW: I think it's easy to get away with this if we document this as an API break, document it properly, make sure some warning flags are raised if someone unintentionally runs into this (i.e. crash right away with a meaningful message). On the other hand I am not very much in favor of trying to get some semi-stable backwards compatibility which will likely never even be used in the wild.

And since we are on this topic, this is the exact reason why I am being so picky with limiting public API to only what is actually useful and stable enough.

elpaso replied:
> Technically you are correct. However, is there a real world use case where the first call to `instance()` is done from outside qgis core? If there is actually such a scenario, and some qgis server plugin does that and subsequently the qgis server core implementation calls `init()` because it has no way of knowing, what would be the desired behavior? I.e. technically speaking isn't that already an API break already?
> 

Speaking of plugins they are all called during request processing and QgsServer is guaranteed to be initialized.

m-kuhn replied:
Is it safe to ignore any use of this class outside plugins? As I understand, it's intended to be used within plugins, and if someone wants to use it outside it's on his own risk?

elpaso replied:
I can't see a use case for that, in any event it is enough to use it after QgsServer is initialized, crashing with a meaningful message is probably a good idea if that condition isn't met.

m-kuhn replied:
It might be self-evident, so please be gentle with me if I'm just stating the obvious.

Crashing is actually a *good thing* because it leaves us with a stack trace at the exact position where something unexpected happens. Just like throwing an exception in python. If we "only" print an error, we may know that something has gone wrong if we actually do have a look at the log and even if we see it, we have no idea *how* we got to this point.

An alternative (in general) would be throwing an exception. In this case here I don't see where and for what reason we should catch that. And it would add considerable complexity (especially if we would want that to traverse to Python).

elpaso replied:
@m-kuhn I think we are on the same page: something like this should work https://github.com/qgis/QGIS/blob/master/src/server/qgsserver.cpp#L74

elpaso commented on src/server/qgsconfigcache.h:
```diff
@@ -65,21 +103,152 @@ class SERVER_EXPORT QgsConfigCache : public QObject
+    //! Remove entry
     void removeChangedEntry( const QString &path );
+
+    //! Remove all cached entries
```
```suggestion
    //! Remove all changed entries
```

dmarteau replied:
Fixed

m-kuhn commented on python/server/auto_generated/qgsconfigcache.sip.in:
```diff
@@ -24,6 +25,11 @@ Cache for server configuration.
 
+    static void initialize( QgsServerSettings *settings );
+%Docstring
+Initialize the strategy
```
Is the term "strategy" defined somewhere in this API doc? It's used twice on this class' doc and impossible to understand without context.

Same for the term `entry` further down.

dmarteau replied:
> Is the term "strategy" defined somewhere in this API doc

This is no longer referenced in the sip from the doc since the method is now out of python scope. Strategy definition is described in  `QgsSettings`.

> Same for the term entry further down.

This what the term used in the original doc, what term should we use ?

m-kuhn commented on src/server/qgsconfigcache.cpp:
```diff
@@ -23,27 +22,73 @@
+    QgsMessageLog::logMessage(
+      QStringLiteral( "Project's cache is already initialized" ),
+      QStringLiteral( "Server" ), Qgis::MessageLevel::Warning );
+    return;
```
Crash instead

dmarteau replied:
Cf discussion: https://github.com/qgis/QGIS/pull/47050#discussion_r799334397

dmarteau replied:
Changed for crashing with Q_ASSERT instead of returning instance with default strategy

pblottiere commented on src/server/qgsserversettings.cpp:
```diff
@@ -615,3 +638,23 @@ QString QgsServerSettings::serviceUrl( const QString &service ) const
+  QString result = value( QgsServerSettingsEnv::QGIS_SERVER_PROJECT_CACHE_STRATEGY ).toString();
+  if ( !result.compare( QLatin1String( "filesystem" ) ) ||
+       !result.compare( QLatin1String( "periodic" ) ) ||
+       !result.compare( QLatin1String( "off" ) ) )
```
`QStringLiteral` instead of `QLatin1String`?

m-kuhn replied:
If there is an overload of `QString::compare( QLatin1String )` the current approach is preferable.

pblottiere replied:
ok thanks :+1:

m-kuhn commented on src/server/qgsserversettings.h:
```diff
@@ -294,6 +296,27 @@ class SERVER_EXPORT QgsServerSettings
+     * - 'filesystem': Use file system watcher for notifying projects change. Note that it works
+     *   only with projects stored in files and not across mounted NFS volumes on Linux.
+     * - 'periodic': Timer based periodic check for project's changes. Works with all storage backend.
+     * - 'off': Disable completely internal project's cache handling
```
Excellent documentation !

m-kuhn commented:
We are getting really close now !

> Is the term "strategy" defined somewhere in this API doc
> 
> This is no longer referenced in the sip from the doc since the method is now out of python scope. Strategy definition is described in QgsSettings.
> 
> Same for the term entry further down.
> 
> This what the term used in the original doc, what term should we use ?

The terms themselves are ok. They are just very general. So for a random reader of this page of the API docs (just looking at this class to find out how things work) he should get an idea where to find more information about this/ what to make of these terms. QgsSettings is probably not the first place to look. And entry could just have a tiny bit more context around (90% of the time, just repeating the name of the method in the doc string is not enough).

dmarteau commented:
> We are getting really close now !

Tests are still missing, they need to be rewrite in as C++ tests.

m-kuhn reviewed (commented):
oops, one point got lost in githubs commenting system

pblottiere commented on src/server/qgsconfigcache.h:
```diff
@@ -30,6 +31,44 @@
+    virtual void entryInserted( const QString &path ) = 0;
+
+    //! Attache cache to this strategy
+    virtual void attach( QgsConfigCache &cache ) = 0;
```
> I have to disagree with @pblottiere on this point, QObject should be passed as pointers, not reference. That's how (almost) all QGIS handles this. And Qt as well.

On the Qt side I don't know but it's true on QGIS side. However I never liked it ^^. I think it's more robust to use reference instead of pointer because we cannot delete it (even by accident).

So what is the technical reason behind this choice?

dmarteau replied:
> Could we just drop this method attach and pass this through the constructor?

Not really, or the initialisation from settings has to be done in the `QgsConfig` cache constructor.

So pointer or not pointer ?

pblottiere replied:
> So pointer or not pointer ?

I think we can blindly trust @m-kuhn of course (no pointer). But I'm just curious to understand why :).

dmarteau replied:
> of course (no pointer) 

Hu, you mean I should revert this to pointer instead reference ?

pblottiere replied:
> Hu, you mean I should revert this to pointer instead reference ?

Yep sorry, it was unclear... -> revert reference and use pointer instead (sorry for the inconvenience)

dmarteau replied:
Reverted to pointer

m-kuhn replied:
> > I have to disagree with @pblottiere on this point, QObject should be passed as pointers, not reference. That's how (almost) all QGIS handles this. And Qt as well.
> 
> On the Qt side I don't know but it's true on QGIS side. However I never liked it ^^. I think it's more robust to use reference instead of pointer because we cannot delete it (even by accident).
> 
> So what is the technical reason behind this choice?

I can only offer "soft" reasons

- When a `QObject` is created it's often used with a `parent` in which case it will be created with `new` and stored in a pointer. So its members are called with `->` so it's strange to change the pattern to a `.` after passing it to a method
- Pointers can also point to `nullptr`
- Qt constructs like `connect` expect pointers
- You shouldn't delete things you don't have ownership of by accident anyway ;-)
- When used as return values, with pointers there is no risk to accidentally take a copy
- Not sure if `QList<QObject &>` works actually?
- Basically I connect the concept of "identity" (you don't copy) to pointer and "value" (you copy whenever you want) to ref/variable (see https://doc.qt.io/qt-5/object.html)

pblottiere replied:
@m-kuhn ok thanks for the explanation :+1:

dmarteau commented:
This PR is still on !!, I missed the time those last weeks.

elpaso commented:
@dmarteau I am happy with the current status of this PR, good work here!

Is there anything left before merge (besides checking the failures in the tests) ?

dmarteau commented:
Tests pushed.  

In order to ease the implementation of tests, I allow creating a standalone `QgsConfigCache`  instance from qgis settings. This does not introduce side effects as `QgsConfigCache` is self contained and does not expose internal implementation.  If this is a problem, it is possible to split the tests in differents independents modules for each strategy.

dmarteau commented:
BTW is think there is a misuse of Qstring::compare in https://github.com/qgis/QGIS/blob/master/src/server/qgsserversettings.cpp#L594. If it is indeed the case,  I may open another PR to fix it.

elpaso commented:
> BTW is think there is a misuse of Qstring::compare in https://github.com/qgis/QGIS/blob/master/src/server/qgsserversettings.cpp#L594. If it is indeed the case, I may open another PR to fix it.

Yeah, I believe it's missing an == 0 or just use plain == if we don't need case insensitive comparison.

Gustry commented on python/server/auto_generated/qgsconfigcache.sip.in:
```diff
@@ -53,10 +61,37 @@ value is ``False``).
+%Docstring
+Returns the name of the current strategy
+
+.. versionadded:: 3.24
```
3.26 ?
For all other `3.24` occurrences in the PR

dmarteau replied:
Yes target is now 3.26

dmarteau replied:
Fixed target version

dmarteau commented:
> 3.26 ?

Yes target is now 3.26

dmarteau commented:
> @dmarteau I am happy with the current status of this PR, good work here!
> 
> Is there anything left before merge (besides checking the failures in the tests) ?

Thx @elpaso.   The tests are covering strategies and settings. I don't see anything left behind: if your ok with it, I think it can be merged.

dmarteau commented:
It appears that the failing test has nothing to do with this PR.

pblottiere commented:
Hello @dmarteau,

I'd still prefer to use enum instead of plain string but Alessandro and Matthias are happy with the current implementation, so +1 to merge as it is! And thanks for having been responsive to all our comments :sweat_smile:.

pblottiere commented:
(the failing test is not related to this PR)

rldhont commented:
Thanks @m-kuhn @elpaso @pblottiere for the reviews. All tests are green.

dmarteau commented:
Is there something needed before merging ?
